Run babel per output

lukastaegert opened the pull request:
Resolves #303.

This PR will add the ability to run Babel on generated chunks instead of input files. Also, this updates all dependencies to their latest versions if possible. I also updated documentation which I will copy below and which should explain most of the details.

One possible improvement that I did not get to yet and that may be part of a future PR is that in order to deduplicate helpers in a code-splitting setup, this plugin could move all used helpers to a separate chunk that is then emitted in the `generateBundle` hook. This should probably not be the default behaviour but might be made configurable via an option. Unfortunately I do not see how it could be possible to only move shared helpers to a shared chunk as we would only know which helpers are shared AFTER all chunks are processed.

I also updated quite a few of the tests as I found them to be either quite unspecific or of the "always green" kind.

Anyway, this is the main part of the documentation update:


## Running Babel on the generated code

By setting `options.transformGenerated: true`, you can configure rollup-plugin-babel to run Babel on the output files instead of the input files. This can be used to perform code transformations on the resulting chunks and is the only way to transform Rollup's auto-generated code. By default, the plugin will be applied to all outputs:

```js
// rollup.config.js
import babel from 'rollup-plugin-babel';
export default {
	input: 'main.js',
	plugins: [babel({ transformGenerated: true })],
	output: [
		{ file: 'bundle.cjs.js', format: 'cjs' },
		{ file: 'bundle.esm.js', format: 'esm' },
	],
};
```

If you only want to apply it to specific outputs, you can use it as an output plugin:

```js
// rollup.config.js
import babel from 'rollup-plugin-babel';
export default {
	input: 'main.js',
	output: [
		{ file: 'bundle.js', format: 'esm' },
		{
			file: 'bundle.min.js',
			format: 'esm',
			plugins: [babel({ transformGenerated: true })],
		},
	],
};
```

The `include`, `exclude` and `extensions` options are ignored when the `transformGenerated` option is used and will produce warnings, and there are a few more points to note that users should be aware of.

### Finding .babelrc files

Depending on what options are used, this plugin will start looking for `.babelrc` files starting in the directory of the file indicated by `output.file` or in the directory indicated by `output.dir`. If neither of these options is used, e.g. when rendering to stdout or when using the JavaScript API, then `.babelrc` files are ignored and options need to be passed in manually.

### Injected helpers

By default, helpers e.g. when transpiling classes will be inserted at the top of each chunk. In contrast to when applying this plugin on the input files, helpers will not be deduplicated across chunks.

You can however use external helpers by adding the `@babel/external-helpers` plugin, which will look for helpers on a global variable named `babelHelpers`:

```js
rollup.rollup({...})
.then(bundle => bundle.generate({
  format: 'esm',
  plugins: [babel({
    transformGenerated: true,
    presets: ['@babel/env'],
    plugins: ['@babel/external-helpers']
  })]
}))
```

```js
// input
export default class Foo {}

// output
var Foo = function Foo() {
	babelHelpers.classCallCheck(this, Foo);
};

export default Foo;
```

Alternatively, you can use imported runtime helpers by adding the `@babel/transform-runtime` plugin. Note that to produce valid code e.g. with `@babel/env`, it is very recommended to configure the format of modules for that plugin to match the output format used by Rollup:

```js
rollup.rollup({...})
.then(bundle => bundle.generate({
  format: 'cjs',
  plugins: [babel({
    transformGenerated: true,
    presets: [['@babel/env', { modules: 'cjs' }]],
    plugins: ['@babel/transform-runtime']
  })]
}))
```

```js
// input
export default class Foo {}

// output
('use strict');

var _interopRequireDefault = require('@babel/runtime/helpers/interopRequireDefault');

var _classCallCheck2 = _interopRequireDefault(require('@babel/runtime/helpers/classCallCheck'));

var Foo = function Foo() {
	(0, _classCallCheck2['default'])(this, Foo);
};

module.exports = Foo;
```

lukastaegert committed:
Update dependencies

lukastaegert committed:
First draft for using the plugin to transform output options

lukastaegert committed:
Deduplicate code

lukastaegert committed:
Pass correct context to transformCode

lukastaegert committed:
Update documentation and add chunking test

lukastaegert committed:
Fix Windows tests

Andarist commented on README.md:
```diff
@@ -112,6 +113,109 @@ plugins: [
+  format: 'cjs',
+  plugins: [babel({
+    transformGenerated: true,
+    presets: [['@babel/env', { modules: 'cjs' }]],
```
I dont believe this is needed - babel shouldnt apply cjs transform when used in combination with Rollup (in such combination this is Rollup’s job)

I would recommend using `useESModules` flag appropriately for transform-runtime in here.

lukastaegert replied:
When I change it the way you recommend, i.e.
```
presets: [[@babel/env'],
plugins: [['@babel/transform-runtime', { useESModules: false }]],
```
in the test `allows transform-runtime to be used instead of bundled helpers for CJS output`, then I get

```js
'use strict';

import _classCallCheck from "@babel/runtime/helpers/classCallCheck";

var Foo = function Foo() {
  _classCallCheck(this, Foo);
};

module.exports = Foo;
```

which looks VERY broken to me.

The problem is, we are running Babel AFTER Rollup has done its job—there will be no more transformations by Rollup at this point. Thus
- The Babel plugin has complete freedom what it does with the output, but
- It needs to preserve import and export semantics

But maybe I should update the Readme to highlight this better: This is not an issue specific to the runtime transform but to any Babel plugin that modifies imports or exports.

Andarist replied:
> It needs to preserve import and export semantics

That's very true, but also those imports should be inserted with that in mind by Babel. So I'm wondering why they were not. I wonder that maybe this is caused by this:
https://github.com/rollup/rollup-plugin-babel/blob/6a99fdf957308f77c320165eb50e0af6b147f081/src/index.js#L25 
and maybe that should be changed based on the generated format in this scenario? This is just a wild guess - it needs more investigation (if you don't have time do to it, I can get to it in a few days, just let me know)

lukastaegert replied:
That should definitely be investigated before proceeding with this. I will definitely not find time to look into this today, maybe end of the week.

Andarist replied:
I've added a commit fixing this - https://github.com/rollup/rollup-plugin-babel/pull/333/commits/8e38908f6bb1c9332ff61b09b65092337744e25c . Please take a loot if this makes sense. cc @loganfsmyth - I would appreciate your review on this one as well.

loganfsmyth replied:
The meaning of `supportsStaticESM` is `this tool knows how to parse ESM`, not "this file is ESM". The issue here is, I'd assume, more related to [`sourceType`](https://babeljs.io/docs/en/options#sourcetype) defaulting to `module`. We may want to consider defaulting `sourceType` to `script` when `format !== "esm`?

lukastaegert replied:
I agree that makes sense. Will play around with this.

lukastaegert replied:
Using `sourceType` provides much better output for the transform-runtime test and also avoids the unnecessary "use strict" in the other test. I am undecided `supportsStaticESM`, though. It does not seem to have adverse effects when we use Babel to transpile ESM to another format, but not adding it will cause some plugins to automatically go for CJS imports, which seems weird to me, especially if you wanted to keep the ESM format. My gut feeling goes towards keeping it if the output format is ESM.

Andarist replied:
With the definition given by @loganfsmyth :
> The meaning of supportsStaticESM is this tool knows how to parse ESM, not "this file is ESM".

when we use babel as output plugin then it seems that the correct thing to state is "this tool doesn't know how to parse ESM" as there won't be any more parsing on the tool's side. It's slightly unsettling to state that when we know that the target format is ESM though

🤷‍♂

loganfsmyth replied:
I should rephrase that if my reading implied that. I'd think of it as meaning that the output of Babel will be consumed by a tool that knowns about esm. When compiling a bundle, it seems like setting it true when when the format is esm makes sense. Otherwise if you set the format the esm in rollup but then tried to run preset-env, it would try to compile the file to cjs by default.

Andarist commented on src/index.js:
```diff
@@ -53,45 +54,56 @@ function createBabelPluginFactory(customCallback = returnObject) {
 
+		if (transformGenerated) {
+			return {
+				name: 'babel-output',
```
Not sure if this should receive a different name. Its still the same plugin - just uses a different strategy for transformation.

lukastaegert replied:
My motivation was to have different names for the case when we apply this plugin twice, once on the input and once on the output. But as we are not using the plugin cache, I guess this will not matter and should not cause warnings. I will change it.

lukastaegert replied:
I guess I will also add an example of using the plugin twice to the README.

Andarist replied:
There is an old PR for using the cache - https://github.com/rollup/rollup-plugin-babel/pull/235 . If this matter for caching I'm not against a different name - could you maybe take a look at that PR to assess what would have to be done to finish it?

lukastaegert replied:
Actually that plugin appears to utilise a "cacheKey", which circumvents the problem of duplicate names. This one needs to be adapted but it could actually help speed up watch mode. Would be worth another look once this is merged. The open remaining question was how to create a good cacheKey that reflects all involved plugins and version numbers. And this is non-trivial as the plugins could change on a per-file basis while a cacheKey would be constant for the plugin.

Andarist replied:
Ok, resolving this comment as out of the scope of this PR. We can circle back to this after this gets merged in.

Andarist commented on src/index.js:
```diff
@@ -53,45 +54,56 @@ function createBabelPluginFactory(customCallback = returnObject) {
+
+				renderStart() {
+					if (pluginOptions.include || pluginOptions.exclude || pluginOptions.extensions) {
+						warnOnce(
```
Would prefer moving this warning out of this render hook, this can be validated just after entering transformGenerated branch of code

lukastaegert replied:
The problem is that I do not have access to `this.warn` at that point, so all I could do is log to the console or throw a real error, which did not seem ideal to me. `renderStart` will be the first hook that runs and would be the canonical place to throw warnings due to configuration issues, at least for output plugins. For input plugins, it would be `buildStart`.

Andarist replied:
Ok, makes sense 👍

Andarist commented on src/index.js:
```diff
@@ -53,45 +54,56 @@ function createBabelPluginFactory(customCallback = returnObject) {
+				renderChunk(code, chunk, options) {
+					return transformCode(
+						code,
+						{ ...babelOptions, filename: options.file || (options.dir && `${options.dir}/placeholder.js`) },
```
In case of the dir option i would use some more distinct filename for the placeholder, smth like `__rollup_babel_placeholder__.js` or similar

Also - couldnt we provide a final filename for entry files?

lukastaegert replied:
> Also - couldnt we provide a final filename for entry files

So you would want to use the entry file as base for where to look for configuration? This does not seem ideal to me as a bundle could contain files from many different places and this could interfere with the situation where you apply the plugin both on the input and the output. Having the output files as base for where we are looking for configuration would allow keeping those two separate.

> In case of the dir option i would use some more distinct filename for the placeholder, smth like __rollup_babel_placeholder__.js

Why, as I understand, this name does not serve any purpose at all except to tell Babel where to start searching for configs. Maybe I will just change it to `_.js` to make clear the name does not matter.

loganfsmyth replied:
I'd lean toward making this option be `null` by default as a opt-in for users. This means we'd expect users to use a `babel.config.js`. It's only `.babelrc` config files that are loaded per-filename.

lukastaegert replied:
Ah, I was not aware about `babel.config.js` files (and also updated the Readme wrongly). Yes, that makes sense. So you suggest we do not pass in the `filename` in any case but allow users to pass in a filename?

loganfsmyth replied:
Great. In all honesty, it may be eve better to not expose the option at all. Perhaps we could see if anyone requests it, to understand why they want it? Most of the time you'd want a filename to
a) Load a specific config via `.babelrc`
b) To allow `ignore`/`only`/`test`/`include`/`exclude` to work

but in the context of a whole-bundle compilation those don't conceptually apply.

lukastaegert replied:
Makes sense. I removed the "dir"/"filename" handling and changed the docs to mention that we only respect project-wide configuration based on the current working directory. It would still be possible to pass in a `filename`, but it is not documented, and I do not see any danger in leaving this open.

Andarist commented on README.md:
```diff
@@ -112,6 +113,109 @@ plugins: [
+
+The `include`, `exclude` and `extensions` options are ignored when the `transformGenerated` option is used and will produce warnings, and there are a few more points to note that users should be aware of.
+
+### Finding .babelrc files
```
For it to be correct u would also have to mention babel.config.js. How we load config files etc should be described in a common section of the docs (rather than in transformGenerated subsection). I think we shouldnt mention specific config files here - just link to Babel docs where all possibilities are described. In here we could only mention that thing about stdout, javascript API etc. 

If this would be described in a common section then in here we should just mention that configs are loaded “for” output files - rather than for input ones.

lukastaegert replied:
Thanks for pointing that out, I will try to improve this.

Andarist commented on test/test.js:
```diff
@@ -139,33 +147,101 @@ describe('rollup-plugin-babel', function() {
-			assert.ok(warnCalled, 'onwarn was never triggered about unresolved imports');
-			assert.ok(!~code.indexOf(HELPERS));
+			assert.deepStrictEqual(warnings, [
+				`'@babel/runtime/helpers/classCallCheck' is imported by samples${path.sep}runtime-helpers${path.sep}main.js, but could not be resolved – treating it as an external dependency`,
```
Couldnt u normalize separator in Rollup? I think many tools do that, seems to simplify stuff when people integrate with it (like sort of in here)

lukastaegert replied:
I guess we should do that at some point, but right now, this would be a massive breaking change that I am very afraid of. There is a lot to consider, and this needs to be communicated well, and we do need to do a lot of testing on Windows. Maybe something to tackle for Rollup@3.

Andarist replied:
Out of curiosity - Rollup@3? 😄I guess this won't happen in the near future 😉

lukastaegert replied:
There is more-or-less a solid set of breaking changes I would love to come in Rollup@2 which I would not want to extend—actually I think it is already too long. Yes, not too soon.

Andarist commented on test/test.js:
```diff
@@ -139,33 +147,101 @@ describe('rollup-plugin-babel', function() {
+
+function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }
+
+var _classCallCheck = _interopDefault(require('@babel/runtime/helpers/esm/classCallCheck'));
```
Shouldnt this test refer to non-esm helpers? Right now it looks exactly the same as next one. Although im looking from mobile so maybe i havent noticed a difference

lukastaegert replied:
Yes, you are right. It was broken (or superfluous?)  before, my changes probably made this a little more apparent. I will remove one.

Andarist commented on test/test.js:
```diff
@@ -347,4 +441,227 @@ describe('rollup-plugin-babel', function() {
+		});
+	});
+
+	describe('when used as an output plugin', () => {
```
Instead of nesting this describe block I think we could just make it top level one (for better readability)

Andarist commented on test/test.js:
```diff
@@ -347,4 +441,227 @@ describe('rollup-plugin-babel', function() {
+			});
+		});
+
+		it('respects a .babelrc file in the same directory when the "dir" option is used', () => {
```
This would be rather confusing for me as a user that using file/dir is supposed to change if we load babel config files. Is loadPartialConfig called conditionally based on that?

lukastaegert replied:
It is not, but otherwise we cannot pass in a filename, and therefore Babel would not know where to look for configuration files. Using `dir` or `filename` seemed reasonable to me, but if you have better ideas, please go ahead.

We could also say that we completely ignore config files when used on the output, but I am not sure everyone would be happy about that.

lukastaegert replied:
If you do not pass in `dir` or `filename`, config files will be ignored, but `loadPartialConfig` will still properly resolve plugins and presets based on `process.cwd()`.

Andarist commented on test/test.js:
```diff
@@ -347,4 +441,227 @@ describe('rollup-plugin-babel', function() {
+					);
+				});
+		});
+	});
```
This probably doesnt differ much but we could also add a test for preserveModules case.

lukastaegert commented:
Thanks for the very detailed review ❤️

lukastaegert committed:
Address review comments

lukastaegert commented:
I tried to address all comments, open for a second review.

lukastaegert committed:
No longer respect "file" and "dir" options and adjust docs

loganfsmyth commented on src/index.js:
```diff
@@ -53,45 +54,50 @@ function createBabelPluginFactory(customCallback = returnObject) {
 			...babelOptions
 		} = unpackOptions(pluginOptions);
 
+		if (transformGenerated) {
```
I'm be curious for @Andarist's thoughts, but I wonder if this might be better exposed much clearly as really a totally different API, e.g.

```
import babel from "rollup-plugin-babel";

const plugin = babel.generated({ ... })
```
since all you do right now is branch the plugin code right up front anyway, and they could still share implementation logic. That would also allow for clearer documentation of the potentially for differing arguments between the two approaches.

lukastaegert replied:
I actually like the idea. Will see if I can work it into the PR.

Andarist replied:
I like it as well better than the previous version, good suggestion @loganfsmyth 👍

loganfsmyth commented on src/index.js:
```diff
@@ -53,45 +54,50 @@ function createBabelPluginFactory(customCallback = returnObject) {
+				},
+
+				renderChunk(code) {
+					return transformCode(code, babelOptions, overrides, customOptions, this);
```
We probably want to automatically wrap this in an IIFE, since Babel's transforms generally assume that their output will be a module. They don't worry about the potential for variable name collisions when things are hoisted to the top-level scope of a file, for instance.

lukastaegert replied:
That depends on what people do but I am sure that for the majority of users, this is not desired as it will just add indirection. This is only interesting if your target is iife anyway and you want to use script tags. There will be no further transformations or hoisting after Babel does its thing.

loganfsmyth replied:
Ah right, I forgot that Rollup has those modes, I haven't actually used it much, just trying to provide insight from the side of Babel. Is `renderChunk` here guaranteed to run before Rollup applies it's UMD or IIFE `format` wrappers, so they'd wrap around any new top-level variables Babel creates?

lukastaegert replied:
No, it is guaranteed to ran AFTER Rollup applies its transformations, so your comment is valid, but it would definitely not apply when targeting esm or cjs. Thus I wonder if we should address it at all, at least for the initial release.

loganfsmyth replied:
Could we have the plugin throw if the format isn't `cjs` or `esm`? My concern is that it's not a problem you're likely to notice until it starts causing really annoying problems that may not be obvious to users of the plugin.

lukastaegert replied:
Throw would mean it is completely unusable even though it produced valid output, that seems like a bad choice for users. We could warn definitely. As I understand, it should theoretically be possible to set the output of Rollup to `esm` and then let `@babel/env` handle the conversion to other formats like `iife`, `amd` or `system`, is that correct?

nicolo-ribaudo replied:
Yes, transforming esm with Babel should work.

loganfsmyth replied:
I really think throwing by default is important. For example, if a user uses a tagged template literal, they'll end up with a `_templateObject` global on their page. If anything else on the page happened to create that first, maybe because they loaded an advertising snippet or something that was also built with this plugin and happened to use a template literal, their site would then fail to load because it would be a duplicate declaration exception.

I would be absolutely terrified to have a application that declared anything in the global scope. It's super dangerous. If we don't throw by default, it would be the default behavior of this plugin.

lukastaegert replied:
True, but it will not break anything for most cases. That is why I would go for a warning instead of an error. Warnings can easily be turned into errors by users if they wish to, but otherwise I see this as needlessly crippling functionality. I see this as similar to how Rollup handles `eval`–you can create a myriad of bugs, but we still allow it and instead go for a warning with a documentation link instead. I think we should do the same here. If you know what you are doing, this would allow for a much broader range of transformations to be applied via Babel in more situations.

loganfsmyth replied:
To take a step back, what exactly is the concern with adding an IIFE in the non-cjs/esm case? The only reason we're even arguing about error vs warn is because you were concerned about that, but an extra iife isn't going to dramatically affect anything, and guarantees this isn't an issue either way. 

If we landed this as-is, this plugin basically invalidate's Rollup's entire `format: "iife"` option by making the output no longer an IIFE, which is entirely counter to user expectations.

> Warnings can easily be turned into errors by users if they wish to

I would argue that no-one will ever do that, so it might as well be silent.

> I see this as needlessly crippling functionality

I really strongly disagree with that. It's not crippling, it's asking Babel to output code that is known to be unsafe to execute in a general context and specifically breaks the guarantees provided by Rollup's `format`.

lukastaegert replied:
> To take a step back, what exactly is the concern with adding an IIFE in the non-cjs/esm case

1. I do not believe Rollup's wrappers will work properly when wrapped in another IIFE, e.g. a `var` will no longer create a global variable.
2. Rollup is chosen for minimal output to achieve a goal. Having an IIFE inside an IIFE is not minimal.

> If we landed this as-is, this plugin basically invalidate's Rollup's entire format: "iife" option by making the output no longer an IIFE

My limited experience with Babel is that without plugins, Babel does zero transformations, so what you said is obviously wrong. This still holds true if I hand-pick my plugins to just do certain transformations. For instance I might want to replace certain constructs with different ones. Or I actually want to modify the wrapper code, and for some reason, I am fluent with Babel so it is my tool of choice. In that case, I could happily live with a warning, but if the developers throw an error instead I would ask "Why on earth!? This is not my issue you are trying to fix!". To my experience, people usually notice warnings, and if there are no issues, what is the problem? I really do not understand why you are obsessed that it has to be an error.

 >I really strongly disagree with that. It's not crippling, it's asking Babel to output code that is known to be unsafe to execute in a general context
It is not known to be unsafe, this depends, as I said, entirely on the transforms you apply. I see it as a big improvement to allow arbitrary transforms because Babel is such a wide-spread AST transformation engine that will enable people to build more powerful stuff on Rollup. I guess for this particular topic, we will remain in disagreement.

loganfsmyth replied:
All of this assumes `format` is not `esm`/`cjs`:

> I do not believe Rollup's wrappers will work properly when wrapped in another IIFE, e.g. a var will no longer create a global variable.

Fair enough, I wasn't sure of Rollup created globals via `var` or undeclared assignment. This is another example where, since many of Babel's plugins assume that a file is a module, they would treat it as reasonable to rename that `var` expecting there to be no external behavioral changes though.

> My limited experience with Babel is that without plugins, Babel does zero transformations, so what you said is obviously wrong. 

That seems a bit harsh. Your own examples in the PR description and README talk about running `@preset/env` on the entire bundle, for instance, and make to recommendation against that in the `iife`/`umd` case. 99% of users will never stop to think of the specifics of the plugins they are running. Any plugin in the future could also change an implementation detail only to suddenly break a user of this patch's bundle by suddenly creating what it expected to be a module-level variable but is actually a global that conflicts with something somewhere.

> This still holds true if I hand-pick my plugins to just do certain transformations.

How are we expecting users to know that plugins are safe to hand-pick? I'd consider anything that creates a global helper function be be an immediate danger to use, but we don't document that and we could always add a helper in the future. I'd consider any `@babel/*` plugin to be unsafe to use on a global script without risk of future breakage.

> For instance I might want to replace certain constructs with different ones. Or I actually want to modify the wrapper code, and for some reason, I am fluent with Babel so it is my tool of choice. In that case, I could happily live with a warning, but if the developers throw an error instead I would ask "Why on earth!? This is not my issue you are trying to fix!". 

If you are someone who defines is fluent in writing a custom Babel transformation, you probably are aware of the pitfalls of transforming a global script, and you'd absolutely have no problem passing an extra option to opt-into global IIFE/UMD processing.

If you're _not_ someone who knows anything about Babel, but sees that this plugin and tries to give it a show, you're honestly saying that you think a warning is better than a clear error saying `Using this plugin to transform a global script with Babel's standard plugins is dangerous. If you are hoping to write your own plugin transform for Rollup's output, pass 'forceEnable:true' opt out of this verification, otherwise we recommend using the per-file transforms.`?

> To my experience, people usually notice warnings, and if there are no issues, what is the problem? I really do not understand why you are obsessed that it has to be an error.

My concern is specifically the "if there are no issues" mentioned in your comment. Running Babel on a full script specifically promotes an approach to Babel complication that is prone to _future_ unexpected issues, even if there are not _present_ issues. That means that someone can easily set it up, do what they need to do, and then get bitten unexpectedly and frustratingly later on. And when they do, it's not particularly satisfying to say "well you ignored that warning", whereas an error makes it clearer.

> It is not known to be unsafe, this depends, as I said, entirely on the transforms you apply. I see it as a big improvement to allow arbitrary transforms because Babel is such a wide-spread AST transformation engine that will enable people to build more powerful stuff on Rollup. I guess for this particular topic, we will remain in disagreement.

Again, your own examples right in the PR and README, have examples of using `preset-env` on the bundle as a whole. That's absolutely unsafe. What if another bundle loads on the page that is built with a different version of Babel and all of the now-global helper functions are overwritten with ones meant for another Babel version? Maybe your company has a general vendor bundle that gets loaded on every page and then once day it upgrades to Babel 8 but apps are still compiled with Babel 7 or something.

I'm not saying this feature isn't useful, I'm saying easily-misused by default is a bad default. Your `eval` example earlier I'd put at a much lower grade annoyance because it's going to be obvious that it is broken when you actually run your code. In this case of this, it'll seem to work until suddenly some random third-party script takes down your whole website because it injected a different version of one of Babel's helpers or something.

Look, if at the end of the day you'd still rather the warning, it's not the end of the world. That said, it's important to then narrow down exactly what the warning says and what plugins people would actually be using on an overall script. I'd be tempted to disable `babel.config.js` loading by default in this case  to further make it clear that any plugins used should be hand-picked for strictly Rollup's final bundle transformation. Any core `@babel/*` plugin should be assumed to potential inject new globals that could then conflict with other page scripts and thus break your app, so we'd want to very clearly say that users shouldn't run Babel's plugins unless they really know what they are doing.

Andarist replied:
> I'd be tempted to disable babel.config.js loading by default in this case to further make it clear that any plugins used should be hand-picked for strictly Rollup's final bundle transformation.

I'm wondering if this shouldn't be the case there - I see a benefit of loading automatically a unified config, but this use case here is very specific and I'm a little bit afraid that for this case applying unified config might also break stuff badly. I'm on a fence here a little bit - hard to tell what are going to be users' expectations. Arguably a mistake should be fairly quickly noticed and we have `configFile: false` to opt-out of it - so maybe it's OK to keep it how it is now.

loganfsmyth replied:
I'd be fine with it being opt-in, but I don't feel super strongly about it. Given that this isn't really a mode we've had before, it's mostly new ground anyway, so perhaps being cautious about defaults is good?

nicolo-ribaudo commented:
Yes, transforming esm with Babel should work.

lukastaegert committed:
Use plugin.generated for output plugin

lukastaegert committed:
Add warning for formats other than ESM or CJS, update docs

lukastaegert commented:
I have updated the code to use `babel.generated` to transform chunks and also added a warning with a recommended fix for formats other than `esm` or `cjs`.

On a side-node, even Babel's UMD wrapper creates unintended globals, in that case `_typeof`, but I believe this is not a big issue is it will likely be the only one and also should not cause conflicts as the code outside the wrapper is not run in strict mode.

nicolo-ribaudo commented:
> On a side-node, even Babel's UMD wrapper creates unintended globals, in that case _typeof, but I believe this is not a big issue is it will likely be the only one and also should not cause conflicts as the code outside the wrapper is not run in strict mode.

This should be already fixed in `master`

lukastaegert committed:
Throw for not recommended formats but add an override option.

lukastaegert commented:
I have changed it so that it throws now for formats other than esm or cjs but added an option to allow users to override this behaviour. I hope everyone is now happy with this.

Andarist commented on README.md:
```diff
@@ -112,6 +114,145 @@ plugins: [
+
+### Finding Babel configuration options
+
+Besides Babel options that are passed directly to this plugin, this plugin will only respect project-wide [Babel configuration files](https://babeljs.io/docs/en/config-files) based on the current working directory when applying it to the generated code.
```
I know we are here in subsection and all, but rather than mentioning "this plugin" in "this plugin will only respect project-wide" I would make it somehow more explicit that this is just about the `.generated` variant.

Andarist commented on test/test.js:
```diff
@@ -139,33 +145,68 @@ describe('rollup-plugin-babel', function() {
+
+function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }
+
+var _classCallCheck = _interopDefault(require('@babel/runtime/helpers/esm/classCallCheck'));
```
this looks broken (even though has no impact on the test), esm helpers are being referenced in cjs output, could we fix this so it doesnt trip any future readers?

Andarist replied:
This test could just use esm output and then it would be fine I think.

Andarist commented:
I've just put some extra, minor, comments - there is still quite a bit of code & comments I need to go through. Hope to get it done tomorrow but for now - I have to go to sleep 😅

lukastaegert committed:
Address review comments

lukastaegert commented:
Thanks! I addressed the comments except for the `supportsStaticESM` one as I agree that needs further investigation. If you can find some time to have a look yourself, that would be great, otherwise I will investigate myself at the end of the week.

Andarist commented on README.md:
```diff
@@ -112,6 +114,145 @@ plugins: [
+
+By default, helpers e.g. when transpiling classes will be inserted at the top of each chunk. In contrast to when applying this plugin on the input files, helpers will not be deduplicated across chunks.
+
+You can however use external helpers by adding the `@babel/external-helpers` plugin, which will look for helpers on a global variable named `babelHelpers`:
```
IMHO nowadays usage of `@babel/external-helpers` (thus global `babelHelpers` as mentioned here) is super rare and should be barely mentioned in the docs - or at least mentioned last. We shouldn't encourage its usage.

lukastaegert replied:
Makes sense, I will reduce it to a side-note.

Andarist committed:
Fixed issue with incorrect caller data being passed to Babel by generated output plugins

Andarist commented on test/test.js:
```diff
@@ -347,4 +396,319 @@ describe('rollup-plugin-babel', function() {
+				presets: [['@babel/env', { modules: 'cjs' }]],
+				plugins: ['@babel/transform-runtime'],
+			},
+			{ format: 'esm' },
```
I think the implementation of this test is somewhat too specific for what it is supposed to test. A combination of `{ modules: 'cjs' }` and `{ format: 'esm' }` looks special - but it has nothing to do with a simple thing of "allows transform-runtime to be used instead of bundled helpers".

I think this could have been written like this because of the problem we've discussed here - https://github.com/rollup/rollup-plugin-babel/pull/333/files#r348753235 , if yes - then it can be updated now as the issue got fixed.

lukastaegert replied:
Updated

Andarist commented on README.md:
```diff
@@ -112,6 +114,145 @@ plugins: [
+
+### Finding Babel configuration options
+
+Besides Babel options that are passed directly to this plugin, `babel-generated(...)` will only respect project-wide [Babel configuration files](https://babeljs.io/docs/en/config-files) based on the current working directory when applying it to the generated code.
```
It would be good to add some justification for this - it might not be obvious why babelrc files are being ignored in this scenario.

lukastaegert replied:
Updated

lukastaegert replied:
Added a note

lukastaegert committed:
Reduce information about external helpers as those are not recommended

lukastaegert committed:
Add `sourceType` when transforming the output.

lukastaegert committed:
Update test and readme for runtime helpers

lukastaegert committed:
Add note about babelrc files

lukastaegert commented:
I addressed most of the comments, except the question if we should ignore project-wide configuration. Still undecided. In any case, we should probably add a note about the `configFile` option so that users know how to override whatever the default is.

Andarist commented:
Ok, so the only thing we need to decide is if we should load configs for output plugins, right? I **think** I would be in favor of not loading them, but I don't feel very strongly about it 🤷‍♂

lukastaegert commented:
By now I also agree not loading by default should be preferable. I will change this by tomorrow and make sure I document the configFile option

lukastaegert committed:
Do not load config files by default

lukastaegert commented:
I have deactivated default config file loading for `babel.generated` and adjusted the documentation to reflects this and also explain the `configFile` option, please have another look.

lukastaegert committed:
Update dependencies

Andarist commented:
Last question - should we bump peerDep on rollup? Can we validate if someone tries to use an output plugin with older version of rollup? Or is note in a doc about supported versions sufficient?

lukastaegert commented:
Unless you want to use it as an output plugin, babel.generated is fully backwards compatible, so I guess a note in the docs would be sufficient.

Andarist commented:
Could you add such? Other than that - I have no further objections. So we should be able to merge this in this week.

cc @loganfsmyth could you maybe take a look as well if everything is OK in your eyes?

lukastaegert committed:
Note minimal Rollup version for output plugins

lukastaegert commented:
I added a note

Andarist commented:
In the following days I'm going to look through PRs and such, need to migrate this plugin to the plugin monorepo. @lukastaegert would you prefer to have this published ASAP or can it wait after I'm done with migration? Note that migration will most likely end up major bumping this package anyway.

rolftimmermans commented:
Is there a way to try this out yet? Latest release on NPM does not have this feature, and using the repository directly does not seem to work because there are no files in `dist`...

Andarist commented:
I hope to release this in a couple of days - maybe today. Stay tuned.

Andarist commented:
I've just published a new `rollup-plugin-babel@next` (5.0.0-alpha.1) with this feature included - please give it a try.

rolftimmermans commented:
It works pretty well for me!

dilyanpalauzov commented:
When will this be released as @current instead of @next?

Andarist commented:
Hard to tell - please track https://github.com/rollup/plugins/pull/108. I really hope to review it soon.

dilyanpalauzov commented:
I do not see a connection between releasing a version offering babel() as ouput plugin and moving to a different repository.

Andarist commented:
The difference is in time put into releasing it and possible maintenance, user issues, etc.

It is currently released as next, so if you care about this feature - just use that next version.
